include/boost/geometry/strategies/transform/matrix_transformers.hpp

```diff
@@ -57,128 +156,100 @@ template
 {
+protected:
+    typedef CalculationType ct;
+    typedef boost::qvm::mat<ct, Dimension2 + 1, Dimension1 + 1> matrix_type;
+public:
+
+    template<typename Matrix>
+    inline matrix_transformer(Matrix const& matrix)
+    {
+        qvm::assign(this->m_matrix, matrix);
+    }
+    inline matrix_transformer() {}
 };
 
 
 template <typename CalculationType>
-class matrix_transformer<CalculationType, 2, 2>
+class matrix_transformer<CalculationType, 2, 2> : public detail::matrix_transformer::matrix_transformer<CalculationType, 2, 2>
 {
 protected :
     typedef CalculationType ct;
     typedef boost::qvm::mat<ct, 3, 3> matrix_type;
-    matrix_type m_matrix;
-
 public :
 
+    template<typename Matrix>
+    inline matrix_transformer(Matrix const& matrix)
+    {
+        qvm::assign(this->m_matrix, matrix);
+    }
+    inline matrix_transformer() {}
     inline matrix_transformer(
                 ct const& m_0_0, ct const& m_0_1, ct const& m_0_2,
                 ct const& m_1_0, ct const& m_1_1, ct const& m_1_2,
                 ct const& m_2_0, ct const& m_2_1, ct const& m_2_2)
     {
-        qvm::A<0,0>(m_matrix) = m_0_0;   qvm::A<0,1>(m_matrix) = m_0_1;   qvm::A<0,2>(m_matrix) = m_0_2;
-        qvm::A<1,0>(m_matrix) = m_1_0;   qvm::A<1,1>(m_matrix) = m_1_1;   qvm::A<1,2>(m_matrix) = m_1_2;
-        qvm::A<2,0>(m_matrix) = m_2_0;   qvm::A<2,1>(m_matrix) = m_2_1;   qvm::A<2,2>(m_matrix) = m_2_2;
-    }
-
-    inline matrix_transformer(matrix_type const& matrix)
-        : m_matrix(matrix)
-    {}
-
-
-    inline matrix_transformer() {}
-
-    template <typename P1, typename P2>
-    inline bool apply(P1 const& p1, P2& p2) const
-    {
-        assert_dimension_greater_equal<P1, 2>();
-        assert_dimension_greater_equal<P2, 2>();
-
-        ct const& c1 = get<0>(p1);
-        ct const& c2 = get<1>(p1);
-
-        ct p2x = c1 * qvm::A<0,0>(m_matrix) + c2 * qvm::A<0,1>(m_matrix) + qvm::A<0,2>(m_matrix);
-        ct p2y = c1 * qvm::A<1,0>(m_matrix) + c2 * qvm::A<1,1>(m_matrix) + qvm::A<1,2>(m_matrix);
-
-        typedef typename geometry::coordinate_type<P2>::type ct2;
-        set<0>(p2, boost::numeric_cast<ct2>(p2x));
-        set<1>(p2, boost::numeric_cast<ct2>(p2y));
-
-        return true;
+        qvm::A<0,0>(this->m_matrix) = m_0_0;   qvm::A<0,1>(this->m_matrix) = m_0_1;   qvm::A<0,2>(this->m_matrix) = m_0_2;
+        qvm::A<1,0>(this->m_matrix) = m_1_0;   qvm::A<1,1>(this->m_matrix) = m_1_1;   qvm::A<1,2>(this->m_matrix) = m_1_2;
+        qvm::A<2,0>(this->m_matrix) = m_2_0;   qvm::A<2,1>(this->m_matrix) = m_2_1;   qvm::A<2,2>(this->m_matrix) = m_2_2;
     }
-
-    matrix_type const& matrix() const { return m_matrix; }
 };
 
 
 // It IS possible to go from 3 to 2 coordinates
 template <typename CalculationType>
-class matrix_transformer<CalculationType, 3, 2> : public matrix_transformer<CalculationType, 2, 2>
+class matrix_transformer<CalculationType, 3, 2> : public detail::matrix_transformer::matrix_transformer<CalculationType, 3, 2>
```

Comment:
So before matrix 3x3 was stored internally, but now it will store 3x4. AFAIU the matrix stored internally should be square matrix representing transformation of the output which in this case is 3x3 since the output is 2d.

Reply:
AFAIU the matrix_transformer is a general affine transformation for 2d to 2d or 3d to 3d, etc., i.e. it represents any composition of a translation and a linear map. For 3d to 2d consider the following transformation: (x, y, z) -> (x + 2z + a, y + 3z + b) represented by the augmented matrix
1 0 2 a
0 1 3 b
0 0 0 1
This is certainly a composition of a linear map (from 3d to 2d) and a translation (2d to 2d) but it cannot be represented as an augmented 3 x 3 matrix, so I think the general case requires (m+1)x(n+1) matrices to represent nd to md transformations.

Besides generality, this has also the advantage that matrix transformers can be composed in the mathematically intuitive way by matrix multiplication, for example, given a 3d to 3d, a 3d to 2d and a 2d to 2d transformation the internal matrices can all be multiplied (in the correct order) and the result can be used to construct a single 3d to 2d transformation that represents the composition of the 3 original transformations.